File: services/market-data/examples/consumer.py
#!/usr/bin/env python3
"""
Example consumer for Market Data Service
Subscribes to order book updates via Redis pub/sub
"""

import redis
import json
import sys
from datetime import datetime

def format_orderbook(data):
    """Format order book for display"""
    symbol = data.get('symbol', 'UNKNOWN')
    timestamp = data.get('timestamp', 0)
    dt = datetime.fromtimestamp(timestamp / 1_000_000)  # microseconds to seconds

    print(f"\n{'='*60}")
    print(f"Symbol: {symbol} | Time: {dt.strftime('%H:%M:%S.%f')}")
    print(f"{'='*60}")

    bids = data.get('bids', {})
    asks = data.get('asks', {})

    # Get top 5 levels
    print("\nAsks (sellers):")
    ask_items = sorted(asks.items(), key=lambda x: float(x[0]), reverse=True)[:5]
    for price, qty in reversed(ask_items):
        print(f"  {float(price):>12,.2f} | {float(qty):>10,.4f}")

    print("\n" + "-"*40)

    print("\nBids (buyers):")
    bid_items = sorted(bids.items(), key=lambda x: float(x[0]), reverse=True)[:5]
    for price, qty in bid_items:
        print(f"  {float(price):>12,.2f} | {float(qty):>10,.4f}")

    # Calculate spread
    if bid_items and ask_items:
        best_bid = float(bid_items[0][0])
        best_ask = float(sorted(asks.items(), key=lambda x: float(x[0]))[0][0])
        spread = best_ask - best_bid
        spread_bps = (spread / best_bid) * 10000
        print(f"\nSpread: ${spread:.2f} ({spread_bps:.2f} bps)")

def format_trade(data):
    """Format trade for display"""
    symbol = data.get('symbol', 'UNKNOWN')
    price = data.get('price', 0)
    quantity = data.get('quantity', 0)
    timestamp = data.get('timestamp', 0)
    is_buyer_maker = data.get('is_buyer_maker', False)
    side = 'SELL' if is_buyer_maker else 'BUY'

    dt = datetime.fromtimestamp(timestamp / 1000)  # milliseconds to seconds

    print(f"[{dt.strftime('%H:%M:%S')}] {symbol} {side:4s} {quantity:>10,.4f} @ ${price:>10,.2f}")

def main():
    symbol = sys.argv[1] if len(sys.argv) > 1 else 'BTCUSDT'

    print(f"Connecting to Redis...")
    r = redis.Redis(host='localhost', port=6379, decode_responses=True)

    # Test connection
    try:
        r.ping()
        print(f"Connected to Redis successfully")
    except redis.ConnectionError:
        print("ERROR: Could not connect to Redis. Is it running?")
        print("Start it with: docker-compose up -d redis")
        sys.exit(1)

    pubsub = r.pubsub()

    # Subscribe to both order book and trades
    orderbook_channel = f'orderbook:{symbol}'
    trade_channel = f'trades:{symbol}'

    print(f"\nSubscribing to:")
    print(f"  - {orderbook_channel}")
    print(f"  - {trade_channel}")
    print(f"\nWaiting for messages... (Ctrl+C to quit)\n")

    pubsub.subscribe(orderbook_channel, trade_channel)

    try:
        for message in pubsub.listen():
            if message['type'] != 'message':
                continue

            channel = message['channel']
            data = json.loads(message['data'])

            if 'orderbook' in channel:
                format_orderbook(data)
            elif 'trades' in channel:
                format_trade(data)

    except KeyboardInterrupt:
        print("\n\nShutting down...")
        pubsub.unsubscribe()
        r.close()

if __name__ == '__main__':
    main()

File: services/market-data/examples/consumer.rs
/// Example consumer for Market Data Service in Rust
/// Subscribes to order book updates via Redis pub/sub
///
/// Usage: cargo run --example consumer [SYMBOL]

use redis::{Commands, ControlFlow};
use serde::Deserialize;
use std::env;
use chrono::{DateTime, Utc};

#[derive(Debug, Deserialize)]
struct OrderBook {
    symbol: String,
    bids: std::collections::BTreeMap<String, f64>,
    asks: std::collections::BTreeMap<String, f64>,
    last_update_id: u64,
    timestamp: i64,
}

#[derive(Debug, Deserialize)]
struct Trade {
    symbol: String,
    trade_id: u64,
    price: f64,
    quantity: f64,
    timestamp: i64,
    is_buyer_maker: bool,
}

fn format_orderbook(book: OrderBook) {
    println!("\n{}", "=".repeat(60));
    let dt = DateTime::from_timestamp_micros(book.timestamp).unwrap_or_default();
    println!("Symbol: {} | Time: {}", book.symbol, dt.format("%H:%M:%S%.6f"));
    println!("{}", "=".repeat(60));

    // Top 5 asks (reversed for display)
    println!("\nAsks (sellers):");
    let mut asks: Vec<_> = book.asks.iter().collect();
    asks.sort_by(|a, b| b.0.cmp(a.0));
    for (price, qty) in asks.iter().take(5).rev() {
        let price_f: f64 = price.parse().unwrap_or(0.0);
        println!("  {:>12.2} | {:>10.4}", price_f, qty);
    }

    println!("\n{}", "-".repeat(40));

    // Top 5 bids
    println!("\nBids (buyers):");
    let mut bids: Vec<_> = book.bids.iter().collect();
    bids.sort_by(|a, b| b.0.cmp(a.0));
    for (price, qty) in bids.iter().take(5) {
        let price_f: f64 = price.parse().unwrap_or(0.0);
        println!("  {:>12.2} | {:>10.4}", price_f, qty);
    }

    // Calculate spread
    if let (Some(best_bid), Some(best_ask)) = (bids.first(), asks.last()) {
        let bid_price: f64 = best_bid.0.parse().unwrap_or(0.0);
        let ask_price: f64 = best_ask.0.parse().unwrap_or(0.0);
        let spread = ask_price - bid_price;
        let spread_bps = (spread / bid_price) * 10000.0;
        println!("\nSpread: ${:.2} ({:.2} bps)", spread, spread_bps);
    }
}

fn format_trade(trade: Trade) {
    let dt = DateTime::from_timestamp_millis(trade.timestamp).unwrap_or_default();
    let side = if trade.is_buyer_maker { "SELL" } else { "BUY" };
    println!(
        "[{}] {} {:4} {:>10.4} @ ${:>10.2}",
        dt.format("%H:%M:%S"),
        trade.symbol,
        side,
        trade.quantity,
        trade.price
    );
}

fn main() -> redis::RedisResult<()> {
    let args: Vec<String> = env::args().collect();
    let symbol = args.get(1).map(|s| s.as_str()).unwrap_or("BTCUSDT");

    println!("Connecting to Redis...");
    let client = redis::Client::open("redis://127.0.0.1/")?;
    let mut con = client.get_connection()?;

    // Test connection
    redis::cmd("PING").query::<String>(&mut con)?;
    println!("Connected to Redis successfully");

    let orderbook_channel = format!("orderbook:{}", symbol);
    let trade_channel = format!("trades:{}", symbol);

    println!("\nSubscribing to:");
    println!("  - {}", orderbook_channel);
    println!("  - {}", trade_channel);
    println!("\nWaiting for messages... (Ctrl+C to quit)\n");

    let mut pubsub = con.as_pubsub();
    pubsub.subscribe(&orderbook_channel)?;
    pubsub.subscribe(&trade_channel)?;

    loop {
        let msg = pubsub.get_message()?;
        let payload: String = msg.get_payload()?;
        let channel: String = msg.get_channel_name();

        if channel.contains("orderbook") {
            if let Ok(book) = serde_json::from_str::<OrderBook>(&payload) {
                format_orderbook(book);
            }
        } else if channel.contains("trades") {
            if let Ok(trade) = serde_json::from_str::<Trade>(&payload) {
                format_trade(trade);
            }
        }
    }
}
